sklearn_numba_dpex/common/random.py

```diff
@@ -292,32 +301,36 @@ def _xoroshiro128pp_jump(states, state_idx):
 
     #     return result
 
-    # Initialization is purely sequential so it will be faster on CPU, if a cpu device
-    # is available make sure to use it.
-    if device is None:
-        device = dpctl.SyclDevice()
-    from_cpu_to_device = False
-    if not device.has_aspect_cpu:
-        try:
-            cpu_device = dpctl.SyclDevice("cpu")
-            from_cpu_to_device = True
-        except dpctl.SyclDeviceCreationError:
-            warnings.warn(
-                "No CPU found, falling back random initiatlization to default device."
-            )
+    @dpex.kernel
+    def init_xoroshiro128pp_states(states, seed):
+        """
+        Use SplitMix64 to generate an xoroshiro128p state from a uint64 seed.
 
-    states = dpt.empty(
-        sh=(n_states, 2),
-        dtype=np.uint64,
-        device=cpu_device if from_cpu_to_device else device,
-    )
-    init_xoroshiro128pp_states[1, 1](states)
+        This ensures that manually set small seeds don't result in a predictable
+        initial sequence from the random number generator.
+        """
+        if n_states < one_idx:
+            return
 
-    if from_cpu_to_device:
-        return states.to_device(device)
+        splitmix64_state = init_const_1 ^ seed[zero_idx]
```

Comment:
Note: I find moving blocks of code around hardens understandability of the logic change: it is hard for reviewers to know what has changed in this file with respect to the scope of this PR due to this diff which does not change the logic it seems.

To me this is the only line whose logic changed (accessing `seed[zero_idx]` instead if `seed`) but I am not entirely sure. Is that the case? Is swapping those blocks really needed here?

Reply:
Unfortunately the fix consists in moving those blocks around to a place where caching can be correctly defined so yes moving the blocks is necessary :-(

Indeed you spotted a difference, seed is now passed as a kernel argument rather than a constant variable to avoid re-triggering the jit each time the seed changes.

Reply:
Fine for me. :)